File: usr/share/openmediavault/mkrrdgraph/plugins.d/drivemon.py
# -*- coding: utf-8 -*-
#
# This file is part of OpenMediaVault.
#
# @license   http://www.gnu.org/licenses/gpl.html GPL Version 3
# @author    Volker Theile <volker.theile@openmediavault.org>
# @copyright Copyright (c) 2009-2019 Volker Theile
#
# OpenMediaVault is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# any later version.
#
# OpenMediaVault is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with OpenMediaVault. If not, see <http://www.gnu.org/licenses/>.
import openmediavault.mkrrdgraph
import openmediavault.subprocess
import openmediavault.rpc

class Plugin(openmediavault.mkrrdgraph.IPlugin):
    def create_graph(self, config):
        drives = openmediavault.rpc.call("Drivemon", "getSettings")['drives']
        for c in drives:
            # http://paletton.com/#uid=33r0-0kwi++bu++hX++++rd++kX
            drivename = '/dev/sd' + c
            config.update(
                {
                    'title_drive': drivename,
                    'config_blue': '#0bb6ff',  # blue
                    'drive_letter': c
                }
            )
            args = []
            # yapf: disable
            # pylint: disable=line-too-long
            args.append('{image_dir}/drivestate-_dev_sd{drive_letter}-{period}.png'.format(**config))
            args.extend(config['defaults'])
            args.extend(['--start', config['start']])
            args.extend(['--title', '"{title_drive}{title_by_period}"'.format(**config)])
            args.append('--slope-mode')
            args.extend(['--lower-limit', '-0.5'])
            args.extend(['--upper-limit', '1.5'])
            args.extend(['--units-exponent', '0'])
            args.append('DEF:stateavg={data_dir}/drivemon-_dev_sd{drive_letter}/gauge-state.rrd:value:LAST'.format(**config))

            args.append('LINE2:stateavg{config_blue}:"active"'.format(**config))
            args.append('GPRINT:stateavg:AVERAGE:"%4.2lf Avg"')
            args.append('GPRINT:stateavg:LAST:"%4.2lf Last\l"')

            args.append('COMMENT:"{last_update}"'.format(**config))
            # yapf: enable
            openmediavault.mkrrdgraph.call_rrdtool_graph(args)
        return 0

File: var/www/openmediavault/js/omv/module/admin/diagnostic/service/plugin/Drivemon.js
/**
 * This file is part of OpenMediaVault.
 *
 * @license   http://www.gnu.org/licenses/gpl.html GPL Version 3
 * @author    Volker Theile <volker.theile@openmediavault.org>
 * @copyright Copyright (c) 2009-2020 Volker Theile
 *
 * OpenMediaVault is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * OpenMediaVault is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with OpenMediaVault. If not, see <http://www.gnu.org/licenses/>.
 */
// require("js/omv/workspace/tab/Panel.js")
// require("js/omv/workspace/panel/Textarea.js")
// require("js/omv/workspace/panel/RrdGraph.js")

Ext.define("OMV.module.admin.diagnostic.service.plugin.Drivemon", {
    extend: "OMV.workspace.tab.Panel",
    alias: "omv.plugin.diagnostic.service.drivemon",
    requires: [
        "OMV.Rpc"
    ],

    title: _("Drivemon"),

    initComponent: function() {
        var me = this;
        me.items = [];
        me.callParent(arguments);

        // Execute RPC to get the information required to add tab panels.
        OMV.Rpc.request({
            callback: function(id, success, response) {
                var drives = response['drives'];
                // Create a RRD graph panel per drive.
                drives = drives.split(',');
                for(var name of drives) {
                    me.add(Ext.create("OMV.workspace.panel.RrdGraph", {
                        title: name,
                        rrdGraphName: "drivestate-" + name.replace(/\//g, "_")
                    }));
                }
            },
            relayErrors: false,
            rpcData: {
                service: "Drivemon",
                method: "getSettings"
            }
        });
    }
});